// rappid/plugins/ui/Navigator/Navigator.mjs
// Navigator
// =========

// 'Navigator' creates a new paper (usually smaller) that helps select what part
// of the diagram is shown (especially for a larger diagram) and provides
// a different way how to zoom.

// Example usage:
//
// var nav = new Navigator({
//     paperScroller: paperScroller,
//     width: 200,
//     height: 200,
// });
//
// nav.$el.appendTo('#navigator');
// nav.render();
import $ from 'jquery';
import { util, dia, mvc } from 'jointjs/src/core.mjs';

export const Navigator = mvc.View.extend({

    className: 'navigator',

    events: {
        'mousedown': 'startAction',
        'touchstart': 'startAction',
        'mousedown .joint-paper': 'scrollTo',
        'touchstart .joint-paper': 'scrollTo'
    },

    documentEvents: {
        'mousemove': 'doAction',
        'touchmove': 'doAction',
        'mouseup': 'stopAction',
        'touchend': 'stopAction'
    },

    options: {
        paperConstructor: dia.Paper,
        paperOptions: {},
        /**
         * @deprecated use zoom instead
         */
        zoomOptions: null,
        zoom: { min: 0.5, max: 2 },
        width: 300,
        height: 200,
        padding: 10
    },

    init: function() {

        if (this.options.zoomOptions) {
            // backward compatibility
            this.options.zoom = util.assign({}, this.options.zoom, this.options.zoomOptions);
        } else if (this.options.zoom) {
            this.options.zoom = util.defaults({}, this.options.zoom, this.constructor.prototype.options.zoom);
        }

        util.bindAll(this, 'updateCurrentView', 'doAction', 'stopAction', 'scrollTo');

        // The updateCurrentView is called everytime paperScroller's scrollbars change
        // or the paper is resized. Resize of the paper is normally also acompanied
        // by a scrollbar change (but doesn't have to be). An event is triggered for
        // the vertical and horizontal scrollbar change. That leads to the updateCurrentView
        // to be called upto 4 times per one paperScroller action. Debouncing the method solves
        // this issue but there is definitely room for improvement.
        // + it solves an issue with wrong target paper position while zooming out a paper with
        // negative x-origin
        this.updateCurrentView = util.debounce(this.updateCurrentView, 0);

        var paperScroller = this.options.paperScroller;
        paperScroller.$el.on(`scroll${this.getEventNamespace()}`, this.updateCurrentView);

        var sourcePaper = this.sourcePaper = paperScroller.options.paper;
        this.listenTo(sourcePaper, 'resize', this.updatePaper);

        this.targetPaper = new this.options.paperConstructor(util.merge({
            model: sourcePaper.model,
            interactive: false,
            frozen: true
        }, this.options.paperOptions));
    },

    render: function() {

        this.targetPaper.$el.appendTo(this.el);
        this.targetPaper.unfreeze();

        this.$currentView = $('<div>').addClass('current-view');

        if (this.options.zoom) {
            var $currentViewControl = $('<div>').addClass('current-view-control');
            this.$currentView.append($currentViewControl);
        }

        this.$el.append(this.$currentView).css({
            width: this.options.width,
            height: this.options.height,
            padding: this.options.padding
        });

        // setting right target paper dimension for the first time.
        this.updatePaper(this.sourcePaper.options.width, this.sourcePaper.options.height);

        return this;
    },

    // Updates the navigator's paper size and transformations
    updatePaper: function(width, height) {

        var sourceOrigin = this.sourcePaper.options.origin;
        var sourceScale = this.sourcePaper.scale();
        var navigatorWidth = this.options.width - 2 * this.options.padding;
        var navigatorHeight = this.options.height - 2 * this.options.padding;

        width /= sourceScale.sx;
        height /= sourceScale.sy;

        var ratio = this.ratio = Math.min(navigatorWidth / width, navigatorHeight / height);
        var ox = sourceOrigin.x * ratio / sourceScale.sx;
        var oy = sourceOrigin.y * ratio / sourceScale.sy;

        width *= ratio;
        height *= ratio;

        this.targetPaper.setDimensions(width, height);
        this.targetPaper.setOrigin(ox, oy);
        this.targetPaper.scale(ratio, ratio);

        this.updateCurrentView();
    },

    // Updates the position and size of the navigator's current view rectangle.
    updateCurrentView: function() {

        var ratio = this.ratio;
        var sourceScale = this.sourcePaper.scale();
        var paperScroller = this.options.paperScroller;
        var topLeftCoordinates = paperScroller.clientToLocalPoint(0, 0);
        var paperPosition = this.targetPaper.$el.position();
        var paperOrigin = this.targetPaper.translate();

        // IE returns translate.ty = NaN when ty = 0. It sets transform attribute to 'translate(tx)'.
        // TODO: handle this in the vectorizer
        paperOrigin.ty = paperOrigin.ty || 0;

        this.currentViewGeometry = {
            top: paperPosition.top + topLeftCoordinates.y * ratio + paperOrigin.ty,
            left: paperPosition.left + topLeftCoordinates.x * ratio + paperOrigin.tx,
            width: paperScroller.$el.innerWidth() * ratio / sourceScale.sx,
            height: paperScroller.$el.innerHeight() * ratio / sourceScale.sy
        };

        this.$currentView.css(this.currentViewGeometry);
    },

    startAction: function(evt) {

        evt = util.normalizeEvent(evt);
        const { clientX, clientY } = evt;

        // click on current-view control starts zooming
        // otherwise paper panning is initiated.
        const action = $(evt.target).hasClass('current-view-control') ? 'zooming' : 'panning';
        const { options, currentViewGeometry, sourcePaper } = this;
        const { paperScroller } = options;

        this.delegateDocumentEvents(null, {
            action,
            startClientX: clientX,
            startClientY: clientY,
            startScrollLeft: paperScroller.el.scrollLeft,
            startScrollTop: paperScroller.el.scrollTop,
            startZoom: paperScroller.zoom(),
            startGeometry: currentViewGeometry,
            startScale: sourcePaper.scale()
        });
    },

    doAction: function(evt) {

        evt = util.normalizeEvent(evt);

        const { clientX, clientY, data } = evt;
        const { sourcePaper, options, ratio } = this;
        const { action, startClientX, startClientY, startScrollLeft, startScrollTop, startZoom, startGeometry, startScale, frameId } = data;
        const { paperScroller, zoom } = options;

        switch (action) {

            case 'panning': {
                const { sx, sy } = sourcePaper.scale();
                const x = (clientX - startClientX) * sx;
                const y = (clientY - startClientY) * sy;
                paperScroller.el.scrollLeft = startScrollLeft + x / ratio;
                paperScroller.el.scrollTop = startScrollTop + y / ratio;
                break;
            }

            case 'zooming': {
                // x / width is the ratio of the original width and the requested width
                const { width } = startGeometry;
                const zoomRatio = 1 + (startClientX - clientX) / width / startScale.sx;
                util.cancelFrame(frameId);
                data.frameId = util.nextFrame(() => {
                    paperScroller.zoom(zoomRatio * startZoom, util.defaults({ absolute: true }, zoom));
                });
                break;
            }
        }
    },

    stopAction: function() {

        this.undelegateDocumentEvents();
    },

    // Scrolls the view to the position determined by the event.
    scrollTo: function(evt) {

        evt = util.normalizeEvent(evt);

        var paperOrigin = this.targetPaper.translate();
        // TODO: see updateCurrentView method
        paperOrigin.ty = paperOrigin.ty || 0;

        var offsetX, offsetY;
        // There is no offsetX/offsetY property in the Firefox event
        if (evt.offsetX === undefined) {
            var targetPaperOffset = this.targetPaper.$el.offset();
            offsetX = evt.pageX - targetPaperOffset.left;
            offsetY = evt.pageY - targetPaperOffset.top;
        } else {
            offsetX = evt.offsetX;
            offsetY = evt.offsetY;
        }

        var cx = (offsetX - paperOrigin.tx) / this.ratio;
        var cy = (offsetY - paperOrigin.ty) / this.ratio;

        this.options.paperScroller.center(cx, cy);
    },

    onRemove: function() {

        this.targetPaper.remove();
        this.options.paperScroller.$el.off(this.getEventNamespace());
    }
});
