// rappid/dist/joint.ui.colorPalette.js
/*! Rappid v3.0.1 - HTML5 Diagramming Framework

Copyright (c) 2015 client IO

 2019-08-10 


This Source Code Form is subject to the terms of the Rappid License
, v. 2.0. If a copy of the Rappid License was not distributed with this
file, You can obtain one at http://jointjs.com/license/rappid_v2.txt
 or from the Rappid archive as was distributed by client IO. See the LICENSE file.*/


this.joint = this.joint || {};
(function (exports, SelectBox_mjs, $) {
    'use strict';

    $ = $ && $.hasOwnProperty('default') ? $['default'] : $;

    var ColorPalette = SelectBox_mjs.SelectBox.extend({

        className: 'select-box color-palette',

        position: function() {

            var $selection = this.$('.select-box-selection');
            var selectionHeight = $selection.outerHeight();
            var selectionOffset = $selection.offset();

            var left = selectionOffset.left;
            var top = selectionOffset.top + selectionHeight;

            if (this.options.target !== document.body) {

                this.$target = this.$target || $(this.options.target);

                // Position relative to target element
                var targetOffset = this.$target.offset();
                left -= targetOffset.left - this.$target.scrollLeft();
                top -= targetOffset.top - this.$target.scrollTop();
            }

            this.$options.css({ left: left, top: top });
        }

    }, {

        // Statics

        OptionsView: SelectBox_mjs.SelectBox.OptionsView.extend({

            renderOptionContent: function(option) {

                var $option = $('<div/>', { 'class': 'select-box-option-content' });

                $option.css('background-color', option.content);

                if (option.icon) {
                    $option.prepend($('<img/>', {
                        'class': 'select-box-option-icon',
                        src: option.icon
                    }));
                }

                return $option;
            }
        })

    });

    exports.ColorPalette = ColorPalette;

}(this.joint.ui = this.joint.ui || {}, joint.ui, $));
