// rappid/dist/joint.ui.popup.js
/*! Rappid v3.0.1 - HTML5 Diagramming Framework

Copyright (c) 2015 client IO

 2019-08-10 


This Source Code Form is subject to the terms of the Rappid License
, v. 2.0. If a copy of the Rappid License was not distributed with this
file, You can obtain one at http://jointjs.com/license/rappid_v2.txt
 or from the Rappid archive as was distributed by client IO. See the LICENSE file.*/


this.joint = this.joint || {};
(function (exports, ContextToolbar_mjs, core_mjs) {
    'use strict';

    // ui.Popup is like ui.ContextToolbar except that it can contain any HTML.

    var Popup = ContextToolbar_mjs.ContextToolbar.extend({

        className: 'popup',

        eventNamespace: 'popup',

        events: {},

        renderContent: function() {

            var content = core_mjs.util.isFunction(this.options.content) ? this.options.content(this.el) : this.options.content;
            if (content) {
                this.$el.html(content);
            }
        }
    });

    exports.Popup = Popup;

}(this.joint.ui = this.joint.ui || {}, joint.ui, joint));
